cpp/src/sampling/detail/sampling_utils_impl.cuh

```diff
@@ -448,5 +626,455 @@ sample_edges(
+      if (edge_ids) {
+        if (edge_types) {
+          if (hops) {
+            thrust::sort_by_key(handle.get_thrust_policy(),
+                                labels->begin(),
+                                labels->end(),
+                                thrust::make_zip_iterator(majors.begin(),
+                                                          minors.begin(),
+                                                          edge_ids->begin(),
+                                                          edge_types->begin(),
+                                                          hops->begin()));
+          } else {
+            thrust::sort_by_key(
+              handle.get_thrust_policy(),
+              labels->begin(),
+              labels->end(),
+              thrust::make_zip_iterator(
+                majors.begin(), minors.begin(), edge_ids->begin(), edge_types->begin()));
+          }
+        } else {
+          if (hops) {
+            thrust::sort_by_key(
+              handle.get_thrust_policy(),
+              labels->begin(),
+              labels->end(),
+              thrust::make_zip_iterator(
+                majors.begin(), minors.begin(), edge_ids->begin(), hops->begin()));
+          } else {
+            thrust::sort_by_key(
+              handle.get_thrust_policy(),
+              labels->begin(),
+              labels->end(),
+              thrust::make_zip_iterator(majors.begin(), minors.begin(), edge_ids->begin()));
+          }
+        }
+      } else {
+        if (edge_types) {
+          if (hops) {
+            thrust::sort_by_key(
+              handle.get_thrust_policy(),
+              labels->begin(),
+              labels->end(),
+              thrust::make_zip_iterator(
+                majors.begin(), minors.begin(), edge_types->begin(), hops->begin()));
+          } else {
+            thrust::sort_by_key(
+              handle.get_thrust_policy(),
+              labels->begin(),
+              labels->end(),
+              thrust::make_zip_iterator(majors.begin(), minors.begin(), edge_types->begin()));
+          }
+        } else {
+          if (hops) {
+            thrust::sort_by_key(
+              handle.get_thrust_policy(),
+              labels->begin(),
+              labels->end(),
+              thrust::make_zip_iterator(majors.begin(), minors.begin(), hops->begin()));
+          } else {
+            thrust::sort_by_key(handle.get_thrust_policy(),
+                                labels->begin(),
+                                labels->end(),
+                                thrust::make_zip_iterator(majors.begin(), minors.begin()));
+          }
+        }
+      }
+    }
+
+    size_t num_unique_labels =
+      thrust::count_if(handle.get_thrust_policy(),
+                       thrust::make_counting_iterator<size_t>(0),
+                       thrust::make_counting_iterator<size_t>(labels->size()),
+                       [d_labels = labels->data()] __device__(size_t idx) {
```

Comment:
You can use predefined `is_first_in_run_t`.
```
        auto num_unique_majors =
          thrust::count_if(rmm::exec_policy(loop_stream),
                           thrust::make_counting_iterator(size_t{0}),
                           thrust::make_counting_iterator(this_chunk_size),
                           is_first_in_run_t<vertex_t const*>{tmp_majors.data()});
```
https://github.com/rapidsai/cugraph/blob/branch-23.04/cpp/src/structure/renumber_edgelist_impl.cuh#L449

Reply:
Thanks for the link. I knew we did something like this in the renumbering code, but I wasn't able to find it.

Refactored in the next push to look like the linked renumbering code.